clippy_lints/src/types/mod.rs

```diff
@@ -188,6 +188,37 @@ declare_clippy_lint! {
     "a borrow of a boxed type"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for usage of [`&Option<T>`](https://doc.rust-lang.org/std/option/index.html) anywhere in the code.
+    ///
+    /// ### Why is this bad?
+    /// An `&Option<T>` parameter prevents calling the function if the caller holds a different type, e.g. `Result<T, E>`.
+    /// Using `Option<&T>` generalizes the function, e.g. allowing to pass `res.ok().as_ref()`
+    /// Returning `&Option<T>` needlessly exposes implementation details and has no advantage over `Option<&T>`.
+    ///
+    /// ### Example
+    /// ```rust,compile_fail
+    /// fn foo(bar: &Option<i32>) -> &Option<i32> { bar }
+    /// fn call_foo(bar: &Result<i32, ()>) {
+    ///     foo(bar.ok()); // does not work
+    /// }
+    /// ```
+    ///
+    /// Use instead:
+    ///
+    /// ```rust
+    /// fn foo(bar: Option<&i32>) -> Option<&i32> { bar }
+    /// fn call_foo(bar: &Result<i32, ()>) {
+    ///     foo(bar.ok().as_ref()); // works!
+    /// }
+    /// ```
```

Comment:
Let's go with a `String` or anything `!Copy` rather than `i32` to demonstrate the issue